### PR TITLE
refactor and increase test coverage

### PR DESCRIPTION
**Pull request by BenReissaus**
also fixes [this](https://github.com/BenReissaus/ESB-DataValidator/issues/8) issue.

### CODE REVIEW

**Comment by BenReissaus** [comment]
Forgot one test case: empty topics! Will update asap.

**Comment by BenReissaus** [comment]
Increased Test coverage. More refactoring will be done when the validator will be made iterative, e.g. using custom Kafka Deserializer/Serializers instead of "StringSerializer".

File: src/main/scala/org/hpi/esb/datavalidator/data/Statistics.scala [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
 
+import scala.util.Try
+
 object Statistics extends Record[Statistics] {
-  def create(stats: String): Statistics = {
+  def deserialize(stats: String): Try[Statistics] = {
     val values = stats.split(",")
-    this (values(0).toLong, values(1).toLong, values(2).toLong, values(3).toLong, values(4).toDouble)
+
+    if (values.length != 5) {
```
Comment (Gnni) [inline src/main/scala/org/hpi/esb/datavalidator/data/Statistics.scala]: Kind of "magic number" - the more queries/data we have, the more of those number we will probably have. So we should might find a way to define that in a more general way. IMHO thats not needed for this PR, but we should discuss about that for future.

File: src/main/scala/org/hpi/esb/datavalidator/validation/ResultValidation.scala [hunk 2]
```diff
@@ -1,7 +1,23 @@
...
 
-abstract class ResultValidation {
-  def execute(records: Records): Boolean
+abstract class ResultValidation(inRecords: ListBuffer[ConsumerRecord[String, String]],
+                                resultRecords: ListBuffer[ConsumerRecord[String, String]]) extends Logging {
+
+  protected def recordsAreIncomplete(): Boolean = {
+    if (inRecords.isEmpty) {
+      logger.info(s"'In' records list is empty. Can not validate correctly.")
+      return true
```
Comment (Gnni) [inline src/main/scala/org/hpi/esb/datavalidator/validation/ResultValidation.scala]: Scalastyle plugin recommends avoiding return statement, so as we are planning to use it for the entire project it might be helpful to consider that already for this PR (thats just an example where I recognised a return).

I tried to convert existing returns in the "one repo project" test project already, so you could have a look at that.

File: src/main/scala/org/hpi/esb/datavalidator/Validator.scala [hunk 3]
```diff
@@ -13,9 +13,11 @@ class Validator extends Configurable with Logging {
...
     val validations = List(
-      new IdentityValidation(config.topics.inTopic, config.topics.outTopic),
-      new StatisticsValidation(config.topics.inTopic, config.topics.statsTopic, config.windowSize)
+      new IdentityValidation(records.getTopicResults(config.topics.inTopic),
+        records.getTopicResults(config.topics.outTopic)),
+      new StatisticsValidation(records.getTopicResults(config.topics.inTopic),
+        records.getTopicResults(config.topics.statsTopic), config.windowSize)
     )
-    validations.foreach(_.execute(records))
+    validations.foreach(_.fulfillsRequirements())
```
Comment (Gnni) [inline src/main/scala/org/hpi/esb/datavalidator/Validator.scala]: It just came into my mind - it might be useful to use _Future_s in future in order to validate queries concurrently, which is not the case with the current implementation if I am right?! Nevertheless, that might not be part of this PR.

File: src/test/scala/org/hpi/esb/datavalidator/data/SimpleRecordTest.scala [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.hpi.esb.datavalidator.data
+
+import org.scalatest.FunSuite
+
+class SimpleRecordTest extends FunSuite {
+  test("testCreate") {
```
Comment (Gnni) [inline src/test/scala/org/hpi/esb/datavalidator/data/SimpleRecordTest.scala]: we should might also test a "negative" case, where it is supposed to now work, i.e., value is not of type long.

File: src/test/scala/org/hpi/esb/datavalidator/data/StatisticsTest.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    val stats = new Statistics()
+    val values = Seq(1,2,3,8,9)
+
+    val newStats = values.foldLeft(stats)((s, value) => s.getUpdatedWithValue(value))
+
+    assert(newStats.min === 1)
+    assert(newStats.max === 9)
+    assert(newStats.avg === 4.6)
+    assert(newStats.sum === 23)
+    assert(newStats.count === 5)
```
Comment (Gnni) [inline src/test/scala/org/hpi/esb/datavalidator/data/StatisticsTest.scala]: values.size? 
scalastyle will probably complain about a lot of magic numbers..

Comment (BenReissaus) [inline src/test/scala/org/hpi/esb/datavalidator/data/StatisticsTest.scala]: I just checked with the scala style guide and in the case of assertions it does not complain ( I would say because the semantics of the values is clear because of the left side). 
It does complain at this code `var timestamps = List(0, 1, 100, 999)` but incorrectly since it's an array assignment and it's clear what the values represent. When I googled I found [this](https://github.com/scalastyle/scalastyle-sbt-plugin/issues/54) issue.

File: src/main/scala/org/hpi/esb/datavalidator/Validator.scala (hunk 3 above)
Comment (BenReissaus) [inline src/main/scala/org/hpi/esb/datavalidator/Validator.scala]: Yes, as soon as all available data for a query validation is available it could be started. We can do that in the future.

File: src/main/scala/org/hpi/esb/datavalidator/data/Statistics.scala (hunk 1 above)
Comment (BenReissaus) [inline src/main/scala/org/hpi/esb/datavalidator/data/Statistics.scala]: I agree. It might be better to have a more explicit serialization/deserialization format such as JSON in which case it would be more clear which value refers to which statistic.

File: src/main/scala/org/hpi/esb/datavalidator/validation/ResultValidation.scala (hunk 2 above)
Comment (BenReissaus) [inline src/main/scala/org/hpi/esb/datavalidator/validation/ResultValidation.scala]: As discussed, you're definitely right about the avoidance of return statements. Since I used them quite a lot and it would involve some more refactoring to have the code more "functional" I would postpone it to later.
